File: src/main/java/net/softbell/bsh/domain/entity/Node.java
package net.softbell.bsh.domain.entity;

import java.io.Serializable;
import java.util.Date;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.NamedQuery;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;


/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 노드 엔티티
 */
@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Entity
@Table(name="node")
@NamedQuery(name="Node.findAll", query="SELECT n FROM Node n")
public class Node implements Serializable
{
	private static final long serialVersionUID = 1L;
	
	@Id
	@GeneratedValue(strategy=GenerationType.IDENTITY)
	@Column(name="node_id", unique=true, nullable=false)
	private int nodeId;

	@Temporal(TemporalType.TIMESTAMP)
	@Column(name="approval_date")
	private Date approvalDate;

	@Column(name="control_mode", nullable=false)
	private byte controlMode;

	@Column(name="enable_status", nullable=false)
	private byte enableStatus;

	@Column(name="node_name", nullable=false, length=50)
	private String nodeName;

	@Temporal(TemporalType.TIMESTAMP)
	@Column(name="register_date", nullable=false)
	private Date registerDate;

	@Column(nullable=false, length=64)
	private String token;

	@Column(nullable=false, length=50)
	private String uid;

	@OneToMany(mappedBy="node")
	private List<NodeConnectionLog> nodeConnectionLogs;

	@OneToMany(mappedBy="node")
	private List<NodeGroupItem> nodeGroupItems;

	@OneToMany(mappedBy="node")
	private List<NodeItem> nodeItems;

	public NodeConnectionLog addNodeConnectionLog(NodeConnectionLog nodeConnectionLog)
	{
		getNodeConnectionLogs().add(nodeConnectionLog);
		nodeConnectionLog.setNode(this);

		return nodeConnectionLog;
	}

	public NodeConnectionLog removeNodeConnectionLog(NodeConnectionLog nodeConnectionLog)
	{
		getNodeConnectionLogs().remove(nodeConnectionLog);
		nodeConnectionLog.setNode(null);

		return nodeConnectionLog;
	}

	public NodeGroupItem addNodeGroupItem(NodeGroupItem nodeGroupItem)
	{
		getNodeGroupItems().add(nodeGroupItem);
		nodeGroupItem.setNode(this);

		return nodeGroupItem;
	}

	public NodeGroupItem removeNodeGroupItem(NodeGroupItem nodeGroupItem)
	{
		getNodeGroupItems().remove(nodeGroupItem);
		nodeGroupItem.setNode(null);

		return nodeGroupItem;
	}

	public NodeItem addNodeItem(NodeItem nodeItem)
	{
		getNodeItems().add(nodeItem);
		nodeItem.setNode(this);

		return nodeItem;
	}

	public NodeItem removeNodeItem(NodeItem nodeItem)
	{
		getNodeItems().remove(nodeItem);
		nodeItem.setNode(null);

		return nodeItem;
	}
}

File: src/main/kotlin/net/softbell/bsh/domain/entity/Node.kt
package net.softbell.bsh.domain.entity

import net.softbell.bsh.domain.EnableStatusRule
import java.io.Serializable
import java.util.*
import javax.persistence.*
import kotlin.collections.ArrayList

/**
 * @author : Bell(bell@softbell.net)
 * @description : 노드 엔티티
 */
@Entity
@Table(name = "node")
@NamedQuery(name = "Node.findAll", query = "SELECT n FROM Node n")
class Node(
        @Temporal(TemporalType.TIMESTAMP)
        @Column(name = "approval_date")
        var approvalDate: Date? = null,

        @Column(name = "control_mode", nullable = false)
        var controlMode: Byte,

        @Column(name = "enable_status", nullable = false)
        var enableStatus: EnableStatusRule,

        @Column(name = "version", nullable = false)
        var version: String,

        @Column(name = "node_name", nullable = false, length = 50)
        var nodeName: String,

        @Column(name = "alias", nullable = false, length = 50)
        var alias: String,

        @Temporal(TemporalType.TIMESTAMP)
        @Column(name = "register_date", nullable = false)
        var registerDate: Date,

        @Column(length = 64)
        var token: String? = null,

        @Column(nullable = false, length = 50)
        var uid: String,

        @OneToMany(mappedBy = "node")
        var nodeConnectionLogs: MutableList<NodeConnectionLog> = ArrayList(),

        @OneToMany(mappedBy = "node")
        var nodeGroupItems: MutableList<NodeGroupItem> = ArrayList(),

        @OneToMany(mappedBy = "node")
        var nodeItems: MutableList<NodeItem> = ArrayList()
) : Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "node_id", unique = true, nullable = false)
    var nodeId: Long = 0

    fun addNodeConnectionLog(nodeConnectionLog: NodeConnectionLog): NodeConnectionLog? {
        nodeConnectionLogs.add(nodeConnectionLog)
        nodeConnectionLog.node = this
        return nodeConnectionLog
    }

//    fun removeNodeConnectionLog(nodeConnectionLog: NodeConnectionLog): NodeConnectionLog? {
//        nodeConnectionLogs.remove(nodeConnectionLog)
//        nodeConnectionLog.node = null
//        return nodeConnectionLog
//    }

    fun addNodeGroupItem(nodeGroupItem: NodeGroupItem): NodeGroupItem? {
        nodeGroupItems.add(nodeGroupItem)
        nodeGroupItem.node = this
        return nodeGroupItem
    }

//    fun removeNodeGroupItem(nodeGroupItem: NodeGroupItem): NodeGroupItem? {
//        nodeGroupItems.remove(nodeGroupItem)
//        nodeGroupItem.node = null
//        return nodeGroupItem
//    }

    fun addNodeItem(nodeItem: NodeItem): NodeItem? {
        nodeItems.add(nodeItem)
        nodeItem.node = this
        return nodeItem
    }

//    fun removeNodeItem(nodeItem: NodeItem): NodeItem? {
//        nodeItems.remove(nodeItem)
//        nodeItem.node = null
//        return nodeItem
//    }

    companion object {
        private const val serialVersionUID = 1L
    }
}